apps/emqx_rule_engine/src/emqx_rule_funcs.erl

```diff
@@ -690,6 +691,16 @@ find_s(S, P, Dir) ->
         SubStr -> SubStr
     end.
 
+jq(FilterProgram, JSONBin)
+  when is_binary(FilterProgram), is_binary(JSONBin) ->
+    case jq:parse(FilterProgram, JSONBin) of
+        {ok, Result} -> erlang:iolist_to_binary(Result);
```

Comment:
nit, I guess just return iolist should be fine. let caller decide what to deal with it.

Reply:
Not sure what is best here. Result is guaranteed to be a list of binaries where each binary contains a valid JSON string. When we are concatenating  everything together like this, the output might not be a valid JSON string. For example, `jq:parse` might return `{ok, [<<"{what:1}">>, <<"{what:2}">>]}` and then we will return <<"{what:1}{what:2}">> which is not a valid JSON string. However, for our use cases, I think we will almost always get a result list with only a single item.

Perhaps a good solution would be to return:

```erlang
case Result of
   [JSONString] -> JSONString;
   ListOfJSONStrings -> erlang:iolist_to_binary([<<"[">>, string:join(ListOfJSONStrings, <<",">>), <<"]">>])
end
```

That way, the output will always be a valid JSON string. @terry-xiaoyu what do you think about the different options?

It looks like the other functions in the same file make sure to return binaries instead of iolists so maybe that is a requirement. @terry-xiaoyu Are strings in the rule engine represented as binaries?

Reply:
I think we should let the `emqx_rule_funcs:jq/2` return `list(jiffy:jiffy_decode_result())`, so the user can use the jq function like this:

```
SELECT
  jq('.a|.[]', '{"a": [{"b": 1}, {"b": 2}, {"b": 3}]}') as jq_output,
  jq_output[1].b as first_b
```

This will get:
```
{"first_b": 1}
```

Reply:
Fixed in [0442bf9](https://github.com/emqx/emqx/pull/7384/commits/0442bf98354a3ff54b7e2e2f5c0bb4a0610f32a9). Please let me know if you think it looks alright.